src/main/kotlin/tools/aqua/konstraints/dsl/SMTFunction.kt

```diff
@@ -233,9 +233,15 @@ fun <T : Sort, S1 : Sort, S2 : Sort, S3 : Sort, S4 : Sort, S5 : Sort> SMTProgram
  */
 class Const<T : Sort>(val sort: T, val program: SMTProgramBuilder, val name: String) {
   operator fun getValue(thisRef: Any?, property: KProperty<*>): Expression<T> {
-    program.registerFun(name, sort, emptyList())
-    return UserDeclaredExpression(name.symbol(), sort)
+    if (const == null) {
+      program.registerFun(name, sort, emptyList())
+      const = UserDeclaredExpression(name.symbol(), sort)
```

Comment:
this involves three locations (context, commands, the returned object) that do not know about each other. can we unify this?

Reply:
also, there is a lot of duplication going on here; can we refactor into a superclass or (inlined?) function?